### PR TITLE
ENH: add OLSVectorized, initial version

### PR DESCRIPTION
**Pull request by josef-pkt**
see #2203 adding special_linear_model
see #4771 adding vectorized OLS

this subclasses OLS and RegressionResults but only a subset of results will work

The core attributes are working correctly, 
Among the methods I only checked conf_int and t_test so far.

### CODE REVIEW

File: statsmodels/regression/special_linear_model.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from statsmodels.regression.linear_model import OLS, RegressionResults
+
+from numpy.testing import assert_allclose
+
+
+class OLSVectorizedResults(RegressionResults):
+    r"""results class for vectorized OLS
+    """
+
+    @cache_writable()
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: Why writable?  This is used in very few places

File: statsmodels/regression/special_linear_model.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        wresid = self.wresid
+        return (wresid * wresid).sum(0)
+
+    @cache_readonly
+    def uncentered_tss(self):
+        wendog = self.model.wendog
+        return (wendog * wendog).sum(0)
+
+    def conf_int(self, alpha=.05, cols=None):
+        #print('using OLSVectorizedResults.conf_int')
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: Delete

File: statsmodels/regression/special_linear_model.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        ci =  super(OLSVectorizedResults, self).conf_int(alpha=alpha,
+                     cols=cols)
+
+        return np.rollaxis(ci, -1)
+
+    def t_test(self, r_matrix, cov_p=None, scale=None, use_t=None):
+        """
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: Is a large part of this docstring copied from somewhere else?  I.e. shareable?

File: statsmodels/regression/special_linear_model.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        -------
+        res : ContrastResults instance
+            The results for the test are attributes of this results instance.
+            The available results have the same elements as the parameter table
+            in `summary()`.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import statsmodels.api as sm
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: If no examples, remove section

File: statsmodels/regression/special_linear_model.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    def summary(self):
+        from statsmodels.iolib.summary import summary_params_2dflat
+        summ = summary_params_2dflat(self,
+                                     endog_names=self.model.endog_names)[1]
+        return summ
+
+
+class OLSVectorized(OLS):
+    _results_class = OLSVectorizedResults
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: Doing this is better than setting it in init, but to make this work you had to put the model class below the results class, contrary to the pattern.  Prettier to follow statespace/regimeswitching precedent and make this a property.

File: statsmodels/regression/tests/test_special_linear_model.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+        attrs = ['params', 'scale', 'bse', 'tvalues', 'pvalues', 'ssr', 'centered_tss',
+                 'rsquared', 'llf', 'aic', 'bic', 'fvalue', 'f_pvalue']
+        for attr in attrs:
+            value = getattr(res1, attr)
+            res2_value = np.array([getattr(r, attr) for r in res2_list]).T
+            assert_allclose(value, res2_value, rtol=1e-13)
+
+    def test_methods(self):
```
Comment (jbrockmendel) [inline statsmodels/regression/tests/test_special_linear_model.py]: Test_confint

File: statsmodels/regression/tests/test_special_linear_model.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        cls.res1 = OLSVectorized(endog, exog).fit()
+        cls.res2_list = [OLS(endog[:, i], exog).fit() for i in range(4)]
+
+    def test_attributes(self):
+        res1 = self.res1
+        res2_list = self.res2_list
+
+
+        attrs = ['params', 'scale', 'bse', 'tvalues', 'pvalues', 'ssr', 'centered_tss',
```
Comment (jbrockmendel) [inline statsmodels/regression/tests/test_special_linear_model.py]: Parametrize

File: statsmodels/regression/tests/test_special_linear_model.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        endog = y_true + sig_e * np.random.randn(*y_true.shape)
+
+        cls.res1 = OLSVectorized(endog, exog).fit()
+        cls.res2_list = [OLS(endog[:, i], exog).fit() for i in range(4)]
+
+    def test_attributes(self):
+        res1 = self.res1
+        res2_list = self.res2_list
+
+
```
Comment (jbrockmendel) [inline statsmodels/regression/tests/test_special_linear_model.py]: Extra line

File: statsmodels/regression/special_linear_model.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Nov 13 16:24:12 2018
+
+Author: Josef Perktold
+"""
```
Comment (jbrockmendel) [inline statsmodels/regression/special_linear_model.py]: Is there a more informative name than “special”?  (or is that standard in a literature I’m not familiar with?)

**Comment by jbrockmendel** [comment]
Big picture, what’s the difference between VectorizedOLS and SUR?

File: statsmodels/regression/special_linear_model.py (hunk 9 above)
Comment (josef-pkt) [reply statsmodels/regression/special_linear_model.py]: There is not a literature common name for it.
This is just "linear models for special cases" (with limited functionality).
However, the intention is to have the models computationally optimized for those special cases.

**Comment by josef-pkt** [comment]
> Big picture, what’s the difference between VectorizedOLS and SUR?

It's a special case with limited functionality that can be vectorized.
General SUR allows for different exog per equation and requires full kronecker or block diagonal long form.
SUR with common exog as in VAR still includes cov_resid and joint inference over all parameters. This special case will be Multivariate OLS as used for MANOVA.

OLSVectorized assumes common exog, but assumes cov_resid is diagonal or irrelevant for inference and only provides inference within an equation but not for the joint params. That is, it is just a computational improvement over running many (univariate endog) OLS in a loop.

**Comment by josef-pkt** [comment]
application residual bootstrap, exog is the same for all bootstrap samples

"The code implements the multiplier bootstrap efficiently by executing all 10,000 regressions simultaneously, exploiting the fact that the regressors are common across the bootstrap replications."

Bruce E. Hansen 2017:  Regression Kink With an Unknown Threshold

bootstrap is used to get simulated p-values that are not pivotal

**Comment by josef-pkt** [comment]
test for joint hypothesis `wald_test` `f_test` seems to be missing

**Comment by zoj613** [comment]
I this still planned for a merge? @josef-pkt 
I am faced with code like the following:
```python
    import statsmodels.api as sm

    n_rows, n_col = some_array.shape
    X = sm.add_constant(temp)
    for j in range(n_cols):
        Y = some_array[:, j]
        model = sm.OLS(Y, X)
        results = model.fit()
       # append results to some  variable
```

This snippet is from a function that is called in another loop many times. It is very slow and I was wondering if there is another way I can speed this loop up?

**Comment by zoj613** [comment]
> @zoj613 Yes, the plan is still to merge this.
> 
> The main open question is what should be in the results class, e.g. vectorized t-test.
> 
> Your loop can be done in a single regression which should be much faster unless there are memory problems with huge X (which would require working in batches). The main issue is what more is in your loop. What statistics do you need from your `results`?

X is a 1d array whose size is equal to the number of rows of `some_array`. How could this be transformed into a single regression? The main statistics I want to extract are: `params`, `rsquared` and `pvalues`.

**Comment by josef-pkt** [comment]
> X is a 1d array

your don't have a constant in your regression?

the basic idea is the
params = pinv(X) @ y
computes the parameters for vectorized regression of each column of y on x.

The main problem is that the residual scale and standard errors are different for each y, which needs some workarounds to get,  for example,  pvalues.

**Comment by josef-pkt** [comment]
If you have only one explanatory variable, then explicit formulas are likely faster. I only looked at multiple regression here.

**Comment by zoj613** [comment]
> @zoj613 Yes, the plan is still to merge this.
> 
> The main open question is what should be in the results class, e.g. vectorized t-test.
> 
> Your loop can be done in a single regression which should be much faster unless there are memory problems with huge ~X~ y (which would require working in batches). The main issue is what more is in your loop. What statistics do you need from your `results`?

Thanks for the suggestion. I was able to get a fast solution by just using the matrix OLS and take advantage of numpy's broadcasting rules. The rsquared and pvalues were fairly easy to compute by hand using scipy's t distribution and formulae for the various sums of squares.

**Comment by josef-pkt** [comment]
rebased
I found my original script when I wrote this.
The example runs, but I'm not sure about the t_test yet.

all main attributes correspond to the loop at rtol 1e-13, I had around with a failure when agreement was around 5e-13.
```
attrs = ['params', 'scale', 'bse', 'tvalues', 'pvalues', 'ssr', 'centered_tss',
         'rsquared', 'llf', 'aic', 'bic', 'fvalue', 'f_pvalue']
```

It should be easy to merge, in case we raise notimplementederror with wald_test and similar.

other methods, e.g. get_prediction ?
predict should work, maybe not get_prediction
we raise in get_influence

**Comment by josef-pkt** [comment]
t_test has unit tests, so only joint hypothesis are not supported
t_test is largely copied code and adjusted for this case

no checks for `predict` yet

no pandas results wrapper. I might skip that here.

summary work but only has the params table.
diaplaying vectorized top table doesn't make sence
We could add a datframe with main results in rows, e.g. rsquared, llf

**Comment by josef-pkt** [comment]
this might just need another review and maybe some small changes before merging.
